File: Other/Java_utils/mpeg2ts/Mpeg2TSSource.java
package com.ibm.mpeg2ts;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.MulticastSocket;
import java.net.SocketAddress;
import java.net.SocketTimeoutException;
import java.util.HashMap;
import java.util.Map;

import org.apache.log4j.Logger;

import com.ibm.frame.Frame;
import com.ibm.frame.PooledQueue;
import com.ibm.mpeg2ts.Mpeg2TSPacket.Mpeg2TSPacketType;
import com.ibm.mpeg2ts.StreamInfo.TS_STREAM_TYPE;

public class Mpeg2TSSource {

	private static Logger log = Logger.getLogger(Mpeg2TSSource.class);

	private static final int MAX_UDP_PACKET_SIZE = 1500;

	private DatagramSocket socket;

	private boolean reTransmitFlag = true;

	private PooledQueue<Frame> videoFrames;

	private int programPID = 0;

	private UdpSource udpSource = null;

	private int outputPort = -1;

	private int socketReceiveBufferSize;

	public Mpeg2TSSource(String ip, int port, PooledQueue<Frame> videoFrames,
			int socketReceiveBufferSize) throws IOException {
		this.socketReceiveBufferSize = socketReceiveBufferSize;
		this.videoFrames = videoFrames;
		socket = createMulticastSocket(new InetSocketAddress(ip, port));
	}

	public Mpeg2TSSource(int port, PooledQueue<Frame> videoFrames,
			int socketReceiveBufferSize, boolean reTransmitFlag)
			throws IOException {
		this.socketReceiveBufferSize = socketReceiveBufferSize;
		outputPort = port + 1;
		this.videoFrames = videoFrames;
		this.reTransmitFlag = reTransmitFlag;
		socket = createUnicastSocket(port);
	}

	public void start() {
		udpSource = new UdpSource(socket, videoFrames, outputPort);
		Thread thread = new Thread(udpSource);
		thread.setName(udpSource.getClass().getSimpleName());
		thread.start();
	}

	public void close() {
		if (udpSource != null) {
			udpSource.close();
		}
	}

	public DatagramSocket createUnicastSocket(int port) throws IOException {

		DatagramSocket socket = null;
		socket = new DatagramSocket(port);
		socket.setReceiveBufferSize(socketReceiveBufferSize);
		socket.setSoTimeout(1000);
		return socket;
	}

	public DatagramSocket createMulticastSocket(SocketAddress address)
			throws IOException {

		DatagramSocket socket = null;

		InetSocketAddress socketAddress = (InetSocketAddress) address;

		if (socketAddress.getAddress().isMulticastAddress()) {
			String osName = System.getProperty("os.name").toLowerCase();
			if (osName.indexOf("linux") != -1) {
				socket = new MulticastSocket(address);
			} else if (osName.indexOf("windows") != -1) {
				socket = new MulticastSocket(socketAddress.getPort());
				socket.setReuseAddress(true);//redundant, already set with empty constructor
			} else {
				log.warn("Subserver is running on unknown O.S. (supporting windows and linux)");
				System.exit(1);
			}

			socket.setReceiveBufferSize(socketReceiveBufferSize);

			((MulticastSocket) socket).joinGroup(socketAddress.getAddress());

		} else {
			socket = new DatagramSocket(address);
		}
		socket.setSoTimeout(1000);
		return socket;
	}

	public class UdpSource implements Runnable {

		private DatagramSocket socket = null;
		private DatagramSocket outputSocket = null;

		private boolean endFlag = false;
		private boolean detectFlag = false;
		private int videoPID;

		private Mpeg2TSParser videoTSParser;
		private Mpeg2TSPacket tsPacket = new Mpeg2TSPacket();

		private PMTFrame pmtFrame = new PMTFrame();
		private Map<Integer, StreamInfo> streamsMap = new HashMap<Integer, StreamInfo>();
		
		private long previousTime = -1;

		DatagramPacket packet = new DatagramPacket(
				new byte[MAX_UDP_PACKET_SIZE], MAX_UDP_PACKET_SIZE);
		DatagramPacket outputPacket = new DatagramPacket(
				new byte[MAX_UDP_PACKET_SIZE], MAX_UDP_PACKET_SIZE);
		private int outputPort;

		private static final int MPEG2TS_PACKET_LENGTH = 188;

		public UdpSource(DatagramSocket socket, PooledQueue<Frame> videoFrames,
				int outputPort) {
			this.socket = socket;
			this.outputPort = outputPort;
			videoTSParser = new Mpeg2TSParser(videoFrames);
		}

		/**
		 * Extracts Mpeg2TSPackets from a given DatagramPacket and sends them to
		 * the corresponding Mpeg2TSParser according to their PID.
		 * 
		 * @param packet
		 *            The packet to extract Mpeg2TSPackets from.
		 */
		private void extractStreams(DatagramPacket packet) {

			for (int offset = 0; offset < packet.getLength(); offset += MPEG2TS_PACKET_LENGTH) {

				tsPacket.fromBytes(packet.getData(), offset, programPID);
				
				if(previousTime != -1) {
					if(detectFlag && ((System.currentTimeMillis() - previousTime) > 1000)) {
						log.info("Timeout detected, now looking for new PID");
						programPID = 0;
						pmtFrame = new PMTFrame();
						detectFlag = false;
					}
				}
				
				if (detectFlag) {				
					if (tsPacket.getPID() == videoPID) {
						previousTime = System.currentTimeMillis();
						videoTSParser.write(tsPacket);
					}
				} else {
					if (tsPacket.getType() == Mpeg2TSPacketType.PAT) {
						programPID = tsPacket.getProgramPID();
					}

					if (tsPacket.getType() == Mpeg2TSPacketType.PMT) { // PMT
																		// received

						if (pmtFrame.addPacket(tsPacket)) {
							// pmtFrame is complete
							streamsMap = pmtFrame.getStreamInfos();

							for (Map.Entry<Integer, StreamInfo> entry : streamsMap
									.entrySet()) {
								StreamInfo info = (StreamInfo) entry.getValue();
								if (info.streamType == TS_STREAM_TYPE.TS_STREAM_TYPE_H264) {
									log.info("New PID detected = " + info.streamPID);
									previousTime = System.currentTimeMillis();
									detectFlag = true;
									videoPID = info.streamPID;
									break;
								}

							}
						}
					}

				}
			}
		}

		@Override
		public void run() {

			boolean firstPacket = true;

			byte[] buffer = new byte[1500];
			packet.setData(buffer);

			if (reTransmitFlag) {
				try {
					outputSocket = new DatagramSocket();
					outputSocket.setSendBufferSize(socketReceiveBufferSize);
					outputPacket.setAddress(InetAddress.getLocalHost());
					outputPacket.setPort(outputPort);
				} catch (Exception e1) {
					e1.printStackTrace();
					log.info("Error creating output socket");
					return;
				}
			}

			while (!endFlag) {

				try {
					//log.info("Before socket.receive ");
					socket.receive(packet);
					//log.info("After socket.receive ");
					if (reTransmitFlag) {
						outputPacket.setData(packet.getData());
						outputPacket.setLength(packet.getLength());
						outputSocket.send(outputPacket);
					}

					if (firstPacket) {
						log.info("At least one packet was received");
						firstPacket = false;
					}

					extractStreams(packet);

				} catch (SocketTimeoutException e) {
					log.info("No packet received");
				} catch (Exception e) {
					log.error(e.toString());
					e.printStackTrace();
				}

			}

			log.info("Before close socket ");
			try {
				cleanSocket();
			} catch (IOException e) {
				log.info("Exception in closing socket - " + e.getMessage());
			}
			log.info("After close socket");

		}

		private void cleanSocket() throws IOException {

			if (outputSocket != null && !outputSocket.isClosed()) {
				outputSocket.close();
				outputSocket = null;
			}
			if (socket != null && !socket.isClosed()) {
				socket.close();
				socket = null;
			}
		}

		public void close() {
			videoTSParser.close();
			endFlag = true;
		}

	}

}

File: work/src/mpeg2ts/Mpeg2TSSource.go
package main

import (
	"net"
	"fmt"
	"time"
	"io/ioutil"

)

const UDP_SIZE = 1500
const FRAME_SIZE = 300000; // 300KB
const MPEG2TS_PACKET_LENGTH = 188
const MAX_UDP_PACKET_SIZE = 1500;


func timeMilisecs() long {
	return long(time.Now().UnixNano()%1e6/1e3)
}

type DatagramPacket struct {
	buf []byte
	length int
}

func NewDatagramPacket (b []byte, l int) *DatagramPacket {return &DatagramPacket{b, l}}
func NewDatagramPacketE () *DatagramPacket {return &DatagramPacket{}}

func (dgPacket* DatagramPacket) GetData() []byte{
	return dgPacket.buf
}

func (dgPacket* DatagramPacket) GetLength() int{
	return dgPacket.length
}

func (dgPacket* DatagramPacket) SetData(data []byte) {
	dgPacket.buf = data
}

type Mpeg2TSSource struct {
	//socket DatagramSocket
	videoFrames frameQueue
    programPID int
	socket net.UDPConn
	//outputPort int
	//socketReceiveBufferSize int
}


//
//type  UdpSource struct {
//	tsPacket Mpeg2TSPacket
//	programPID int // = 0
//	endFlag bool //= false;
//	detectFlag bool //= false;
//	previousTime int64 // = -1
//	pmtFrame PMTFrame //
//}
type UdpSource struct {
	Mpeg2TSSource

	//outputSocket int //DatagramSocket

	endFlag bool
	detectFlag bool
	videoPID int

	videoTSParser Mpeg2TSParser
	tsPacket Mpeg2TSPacket

	pmtFrame PMTFrame
	previousTime long
	streamsMap map[int]StreamInfo //HashMap<Integer, StreamInfo>
	packet DatagramPacket

}

func (u* UdpSource) extractStreams(packet DatagramPacket) {
	for offset := 0; offset < packet.GetLength(); offset += MPEG2TS_PACKET_LENGTH {

		u.tsPacket.FromBytes(packet.GetData(), offset, u.programPID);

		if u.previousTime != -1 {
			if(u.detectFlag && ((timeMilisecs() - u.previousTime) > 1000)) {
				//log.info("Timeout detected, now looking for new PID");
				u.programPID = 0;
				u.pmtFrame = *NewPMTFrame();
				u.detectFlag = false;
			}
		}

		if (u.detectFlag) {
			if (u.tsPacket.GetPID() == u.videoPID) {
				u.previousTime = timeMilisecs();
				u.videoTSParser.Write(u.tsPacket);
			}
		} else {
			if (u.tsPacket.GetType() == Mpeg2TSPacketType_PAT) {
				fmt.Println("PAT found")
				u.programPID = u.tsPacket.GetProgramPID();
			}

			if (u.tsPacket.GetType() == Mpeg2TSPacketType_PMT) { // PMT received
				fmt.Println("PMT received")
				if (u.pmtFrame.AddPacket(u.tsPacket)) {// pmtFrame is complete

					u.streamsMap = u.pmtFrame.GetStreamInfos();

					//for (map.Entry<Integer, StreamInfo> entry : streamsMap
					//.entrySet()) {
					for _, v := range u.streamsMap {
						var info StreamInfo = v
						if (info.streamType == TS_STREAM_TYPE_H264) {
							fmt.Println("New PID detected = ", info.streamPID)
							u.previousTime = timeMilisecs()
							u.detectFlag = true;
							u.videoPID = info.streamPID;
							break;
						}

					}
				}
			}

		}
	}
}

func (u* UdpSource) FrameQueueFiller() {
	var firstPacket bool = true
	var buffer= make([]byte, 1500)
	u.packet.SetData(buffer)
	fmt.Println("Entered FrameQueueFiller")

	for u.endFlag != true {
		rlen, _, err := u.socket.ReadFromUDP(buffer)
		if err != nil {
			fmt.Println("No packet received")
			continue
		}
		if firstPacket {
			fmt.Println("At least one packet was received")
			firstPacket = false
		}
		var pgPacket = NewDatagramPacket(buffer, rlen)
		//fmt.Println(buffer[:30])
		u.extractStreams(*pgPacket)
	}
	fmt.Println("exited loop")
	defer u.socket.Close();
}


func NewMpeg2TSSource(port string ,queue frameQueue) *Mpeg2TSSource {
	addr, _ := net.ResolveUDPAddr("udp", ":"+ port)
	sock, _ := net.ListenUDP("udp", addr)
	return &Mpeg2TSSource{queue, 0, *sock}
}
func NewUdpSource(port string,queue frameQueue) * UdpSource{
	return &UdpSource{
		*NewMpeg2TSSource(port,queue),
		false,
		false,
		0,
		*NewMpeg2TSParser(queue),
		*NewMpeg2TSPacket(),
		*NewPMTFrame(),
		-1,
		make(map[int]StreamInfo),
		*NewDatagramPacketE(),
	}
}


//-----------------------------------------------------------//
// tester.go file code moved here
//-----------------------------------------------------------//

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func WriteFile(buffer []byte, file string){
	err := ioutil.WriteFile(file, buffer, 0644)
	check(err)
}

func CheckIfIFrame(buffer []byte, offset int, length int) bool {
	var startOffset int = offset
	for i := startOffset; i < (length -7); i++ {
		var j int = offset + i
		if buffer[j] == 0 {
			if buffer[j+1] == 0 {
				if buffer[j+2] == 1 {
					var val short= (short) (0x001f & buffer[j + 3])
					if val == 7 {
						return true // Start GOP was detected
					}
				}
			}
		}
	}
	return false
}

func FrameQueueDispatcherFullFile(videoFrames frameQueue) {
	fmt.Println("entered  FULL FILE FrameQueueDispatcher")
	h264Buffer := make([]byte, 60*1024*1024)
	var length int = 0
	var counter = 0
	var iframeDetected bool = false
	for {
		//fmt.Println("entered consmer loop")
		if videoFrames.IsEmpty() {
			fmt.Println("consume sleep 100ms")
			time.Sleep(100 * time.Millisecond)
			if videoFrames.IsEmpty() {
				fmt.Println("consume sleep 8000ms")
				time.Sleep(8000 * time.Millisecond)
				if videoFrames.IsEmpty() {
					break
				}
			}
		}
		var frame *Frame  = videoFrames.Poll();
		if !iframeDetected {
			if CheckIfIFrame(frame.GetData(),0, frame.Size()) {
				iframeDetected = true
			} else {
				videoFrames.Recylce(frame)
				continue
			}
		}

		//we have alot 0 sized p_frames...dunno why, but lets skip this for now:
		if (frame.Size() == 0) {
			videoFrames.Recylce(frame)
			continue
		}
		counter++
		fmt.Println("counter: ",counter)
		ArrayCopy(frame.GetData(),0, h264Buffer,length,frame.Size())
		length += frame.Size()
		videoFrames.Recylce(frame)
		if (counter >= 500){
			break;
		}
	}
	fmt.Println("left FrameQueueDispatcher loop.")
	WriteFile(h264Buffer[:length],"../../full.264");
}




func FrameQueueDispatcher(videoFrames frameQueue) {
	fmt.Println("entered FrameQueueDispatcher")
	//Used to identify frames when written to files.
	var counter = 0
	var i_counter = 0
	var p_counter = 0
	var p_type = 'P'
	var i_type = 'I'
	var s string //= fmt.Sprintf("../tmp/frame[%d]-I[%d]-P[%d]_TYPE<%c>", counter,i_counter,p_counter,p_type)
	var iframeDetected bool = false

	for {
		//fmt.Println("entered consmer loop")
		if videoFrames.IsEmpty() {
			fmt.Println("consume sleep 100ms")
			time.Sleep(100 * time.Millisecond)
			if videoFrames.IsEmpty() {
				fmt.Println("consume sleep 8000ms")
				time.Sleep(8000 * time.Millisecond)
				if videoFrames.IsEmpty() {
					break
				}
			}
		}
		var frame *Frame  = videoFrames.Poll();
		if !iframeDetected {
			if CheckIfIFrame(frame.GetData(),0, frame.Size()) {
				iframeDetected = true
			} else {
				videoFrames.Recylce(frame)
				continue
			}
		}

		//we have alot 0 sized p_frames...dunno why, but lets skip this for now:
		if (frame.Size() == 0) {
			videoFrames.Recylce(frame)
			continue
		}

		counter++
		if CheckIfIFrame(frame.GetData(),0, frame.Size()) {
			i_counter++
			p_counter = 0
			s = fmt.Sprintf("../../tmp/frame[%d]-I[%d]-P[%d]_TYPE(%c)____size_%d-----", counter,i_counter,p_counter,i_type,frame.Size())
		} else {
			p_counter++;
			s = fmt.Sprintf("../../tmp/frame[%d]-I[%d]-P[%d]_TYPE(%c)____size_%d", counter,i_counter,p_counter,p_type,frame.Size())
		}

		fmt.Println("file: ",counter)
		actualData := frame.GetData()[:frame.Size()]
		WriteFile(actualData,s)
		WriteFile(actualData,fmt.Sprint("../../media/",counter))
		videoFrames.Recylce(frame)
		if (counter >= 500){
			break;
		}

	}
	fmt.Println("left FrameQueueDispatcher loop.")
}



